### PR TITLE
Blog tidy collaborative filtering

### PR DESCRIPTION
**Pull request by christophscheuch**
The idea of this blog post is to demonstrate how easily different collaborative filtering algorithms can be implemented in a tidy manner. This demonstration is potentially more interesting for an industry audience than academic.

### CODE REVIEW

File: blog/tidy-collaborative-filtering/thumbnail.png
Comment (christophscheuch) [inline blog/tidy-collaborative-filtering/thumbnail.png:1]: Created by DALL-E 2 btw lol with the prompt "Generate a thumbnail image showcasing two stylized AI robots sitting at a round table, immersed in a discussion with a large, abstract, holographic projection of a collaborative filtering algorithm matrix in the middle. The room has a futuristic setting, with a backdrop of a digital, grid-like landscape symbolizing the internet. The color palette should be a combination of cool blues and warm yell"

File: blog/tidy-collaborative-filtering/index.qmd [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+# A Primer on Recommender Systems
+
+## Collaborative Filtering 
+
+In collaborative filtering, recommendations are purely based on past interactions between users and items to produce new recommendations. The main notion is that past user-item interactions are sufficient to detect similar users or items. Broadly speaking, there are two sub-classes of collaborative filtering: the memory-based approach which essentially searches nearest neighbors based on recorded transactions and is hence model-free; and the model-based approach where new representations of users and items are built based on some generative pre-estimated model. Theoretically speaking, the memory-based approach has a low bias (since there is no latent model assumed), but a high variance (since the recommendations change a lot in the nearest neighbor search). The model-based approach relies on a trained interaction model and has hence a relatively higher bias, but a lower variance, i.e., recommendations are more stable since they come out of a model.
```
Comment (voigtstefan) [inline blog/tidy-collaborative-filtering/index.qmd:21]: Output from Grammarly: In collaborative filtering, recommendations are based on past user interactions and items to produce new recommendations. The central notion is that past user-item interactions are sufficient to detect similar users or items. Broadly speaking, there are two sub-classes of collaborative filtering: the memory-based approach, which essentially searches nearest neighbors based on recorded transactions and is hence model-free, and the model-based approach, where new representations of users and items are built based on some generative pre-estimated model. Theoretically, the memory-based approach has a low bias (since no latent model is assumed) but a high variance (since the recommendations change a lot in the nearest neighbor search). The model-based approach relies on a trained interaction model. It has a relatively higher bias but a lower variance, i.e., recommendations are more stable since they come from a model.

File: blog/tidy-collaborative-filtering/index.qmd [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+wikifolio_portfolios <- read_rds("../data/wikifolio_portfolios.rds")
+```
+
+I load a dataset with stock holdings of investable wikifolios at the beginning of 2023. Reach out if you want to get access to the data. 
+
+```{r}
+glimpse(wikifolio_portfolios)
+```
+
+First, I convert the long data to a binary rating matrix where only non-NA values are stored explicitly and NA values are represented by dots for efficiency.
```
Comment (voigtstefan) [inline blog/tidy-collaborative-filtering/index.qmd:54]: I don't believe that NAs are represented by dots is solving anything. Under the hood, much more is happening to store sparse matrices with the arules package.

File: blog/tidy-collaborative-filtering/index.qmd [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+              names_from = stock,
+              values_from = in_portfolio,
+              values_fill = list(in_portfolio = 0)) |>
+  select(-wikifolio) |>
+  as.matrix() |>
+  as("binaryRatingMatrix")
+binary_rating_matrix
+```
+
+As in our book chapter on factor selection via machine learning, I perform cross-validation and split the data into training and test data. The training sample constitute 80% of the data and I perform $k$-times recursive estimation of the cross validation with $k=5$. Note that the parameter `given=-1` means that all but one randomly selected item is ignored for the evaluation.
```
Comment (voigtstefan) [inline blog/tidy-collaborative-filtering/index.qmd:68]: - [factor selection via machine learning](https://www.tidy-finance.org/factor-selection-via-machine-learning.html)
- "perform recursive estimation of the cross-validation" sounds a bit odd. "I perform 5-fold cross-validation" would be the common way to express this. 
- You probably want to highlight what the hyperparameters are that are trained, right now it is a bit unclear to me what it actually is that is chosen.

File: blog/tidy-collaborative-filtering/index.qmd [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+* Random Items: the benchmark case because it just stupidly chooses random stocks from all possible choices.
+* Popular Items: just recommended the most popular stocks to measured by the number of wikifolios that hold the stock.
+* Association Rules: each wikifolio and its portfolio is considered as a transaction. Association rule mining finds similar portfolios across all traders (if a traders has x, y and z in his/her portfolio, then he/she is X% likely of also including w). 
+* Item-Based Filtering: the algorithm calculates a similarity matrix across stocks. Recommendations are then based on the list of most similar stocks to the ones the wikifolio already has in its portfolio. 
+* User-Based Filtering: the algorithm finds a neighborhood of similar wikifolios for each wikifolio (for this exercise it is set to 100 most similar wikifolios). Recommendations are then based on what the most similar wikifolios have in their portfolio. 
+
+For each algorithm, I base the evaluation on 1, 3, 5, and 10 recommendations. Note that the evaluation takes a couple of hours, in particular because IBCF and UBCF are quite time-consuming.
```
Comment (voigtstefan) [inline blog/tidy-collaborative-filtering/index.qmd:89]: Item-Based and User-Based Filtering are quite time-consuming - no need for abbreviations.

File: blog/tidy-collaborative-filtering/index.qmd [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+```
+
+```{r include = FALSE}
+results <- read_rds("../data/results.rds")
+```
+
+# Evaluate Recommenders
+
+The output of `evaluate()` already provides evaluation metrics in a structured way. I can simply average the metrics over the cross-validation folds.
+
```
Comment (voigtstefan) [inline blog/tidy-collaborative-filtering/index.qmd:120]: Häh, was evaluierst man jetzt? Du hast jetzt jedem wiki ein paar assets angeboten - wie siehst du, ob das auch sinnvoll war?

File: blog/tidy-collaborative-filtering/index.qmd [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+title: "Tidy Collaborative Filtering: Building A Stock Recommender"
+author: "Christoph Scheuch"
+date: "2023-05-14"
+description: A simple implementation for prototyping multiple collaborative filtering algorithms.
+image: thumbnail.png
+categories: 
+  - Recommender System
+---
+
+Recommender systems are a key component of our digital lifes, ranging from e-commerce, online advertisements, movie recommendations, or more generally all kinds of product recommendations. A recommender system aims to efficiently deliver relevant information to users by automatically searching through a large volume of dynamically generated information to provide users with personalized content. In this blog post, I illustrate the concept of recommender systems in building a simple stock recommendation tool that relies on publicly available portfolios from the social trading platform wikifolio.com. The resulting recommender proposes stocks for investors who already have their own portfolio and look for new investment opportunities. The underlying assumption is that the wikifolio traders hold stock portfolios that can provide useful inspiration for other investors.
```
[the remark this answers is not in the document]
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:11]: Rewritten to "A recommender system aims to efficiently deliver personalized content to users out of a large pool of potentially relevant information."

File: blog/tidy-collaborative-filtering/index.qmd (hunk 1 above)
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:21]: Thanks for the suggestion. I checked most paragraphs with Grammarly now.

File: blog/tidy-collaborative-filtering/index.qmd (hunk 2 above)
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:54]: I don't remember what I wanted to say here when I wrote this. Removed "where only non-NA values are stored explicitly and NA values are represented by dots for efficiency"

File: blog/tidy-collaborative-filtering/index.qmd (hunk 3 above)
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:68]: I improved the description of the testing procedure. I don't think that this is the right location to talk about hyperparameters because they depend on the algorithm (if there are any).

File: blog/tidy-collaborative-filtering/index.qmd [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+library(scales)
+library(recommenderlab)
+```
+
+```{r, include = FALSE}
+wikifolio_portfolios <- read_rds("../data/wikifolio_portfolios.rds")
+```
+
+I load a dataset with stock holdings of investable wikifolios at the beginning of 2023. Reach out if you want to get access to the data. 
+
```
[the remark this answers is not in the document]
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:49]: I don't have a good data hosting solution that I can guarantee for the next 12 months. Maybe we create a repository?

File: blog/tidy-collaborative-filtering/index.qmd (hunk 5 above)
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:120]: Habs versucht an 2 Stellen nochmal klarer zu machen: 

Testing is performed by withholding items from the test portfolios (parameter `given`) and checking how well the algorithm predicts the withheld items. The `given=-1` means that an algorithm sees all but 1 withheld stock for the test portfolios and needs to predict the missing stock.

For each algorithm, I base the evaluation on 1, 3, 5, and 10 recommendations. This specification means that each algorithm proposes 1 to 10 recommendations to the test portfolios and the evaluation scheme then checks whether the proposals contain the one withheld stock.

File: blog/tidy-collaborative-filtering/index.qmd (hunk 7 above)
[the remark this answers is not in the document]
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:50]: I decided to wait for either (i) people reach out that they want to have access to the data or (ii) we have found a way to host tidy finance data other than me providing some link to a private cloud data store.

**Review by voigtstefan — APPROVED** [review]

File: blog/tidy-collaborative-filtering/index.qmd [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    train  = 0.8,
+    given  = -1
+)
+scheme
+```
+
+Here is the list of recommenders that I consider for the backtest with some intuition:
+
+* Random Items: the benchmark case because it just stupidly chooses random stocks from all possible choices.
+* Popular Items: just recommended the most popular stocks to measured by the number of wikifolios that hold the stock.
```
Comment (patrick-weiss) [inline blog/tidy-collaborative-filtering/index.qmd:83]: **Typo**: "just recommends the most popular stocks measured by the number of wikifolios that hold the stock."

File: blog/tidy-collaborative-filtering/index.qmd [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+* Association Rules: each wikifolio and its portfolio is considered as a transaction. Association rule mining finds similar portfolios across all traders (if a traders has x, y and z in his/her portfolio, then he/she is X% likely of also including w).
```
Comment (patrick-weiss) [inline blog/tidy-collaborative-filtering/index.qmd:84]: Should we use "they/them" as a neutral choice for including non-binary?

Also some typos "(if a trader has x, y, and z in their portfolio, then they are X% likely also to include w)"

File: blog/tidy-collaborative-filtering/index.qmd [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+The output of `evaluate()` already provides the evaluation metrics in a structured way. I can simply average the metrics over the cross-validation folds.
+
+```{r}
+results_tbl <- results |>
+  avg() |>
+  map(as_tibble) |>
+  bind_rows(.id = "model")
+```
+
+Now, for each recommender, we are interested the following numbers:
```
Comment (patrick-weiss) [inline blog/tidy-collaborative-filtering/index.qmd:128]: **Typo**: "interested in the following number:"

File: blog/tidy-collaborative-filtering/index.qmd [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+* False negative (FN) = number of not predicted items that correspond to withheld items
+* True Positive (TP) = number of correct predictions that correspond to withheld items
+
+The two figures below present the most common evaluation techniques for the performance of recommender algorithms in backtest settings like mine. 
+
+## ROC curves
+
+The first visualization approach comes from signal-detection and is called "Receiver Operating Characteristic" (ROC). The ROC-curve plots the algorithm’s probability of detection (TPR) against the probability of false alarm (FPR).
+
+* TPR = TP / (TP + FN) (i.e. share of true positive recommendations relative to all known portfolios)
```
Comment (patrick-weiss) [inline blog/tidy-collaborative-filtering/index.qmd:141]: **Comma** after i.e.: i.e.,

File: blog/tidy-collaborative-filtering/index.qmd [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+    y = "True Positive Rate (TPR)",
+    title = "ROC curves of stock recommender algorithms",
+    colour = "Model"
+  ) +
+  theme(legend.position = "right") + 
+  scale_y_continuous(labels = percent) +
+  scale_x_continuous(labels = percent)
+```
+
+The figure shows that recommending random items exhibits the lowest TPR for any FPR, so it is the worst among all algorithms (which it is not surprising). Association rules, on the other hand, constitutes the best algorithm among the current selection. This result is neat because association rule mining is a computationally cheap algorithm, so we could potentially fine-tune or reestimate the model easily.
```
Comment (patrick-weiss) [inline blog/tidy-collaborative-filtering/index.qmd:163]: **Typos**:
"(which is not surprising)"
"Association rules, on the other hand, constitute"

File: blog/tidy-collaborative-filtering/index.qmd [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+
+## Precision-Recall Curves
+
+The second popular approach is to plot Precision-Recall curves. The two measures are often used in information retrieval problems:
+
+* Precision = TP / (TP + FP) (i.e. correctly recommended items relative to total recommended items)
```
Comment (patrick-weiss) [inline blog/tidy-collaborative-filtering/index.qmd:169]: i.e.,

File: blog/tidy-collaborative-filtering/index.qmd [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+# Create Predictions
+
+The final step is to create stock recommendations for investors who already have portfolios. I pick the IBCF algorithm to create such recommendations because it excelled in the analyses above. Note that in the case of association rules, I also need to provide the support and confidence parameters. 
+
+```{r}
+recommender <- Recommender(binary_rating_matrix, method = "AR", param = list(supp = 0.01, conf = 0.1))
+```
+
+As ane example, suppose you currently have a portfolio that consists of Nvidia (US67066G1040) and Apple (US0378331005). I have to transform this sample portfolio into a rating matrix with the same dimensions as the data we used as input for our training. The `predict()` function then delivers a prediction for the example portfolio.
```
Comment (patrick-weiss) [inline blog/tidy-collaborative-filtering/index.qmd:200]: **Typo**: "As an example"

File: _quarto.yml [hunk 15]
```diff
@@ -6,6 +6,7 @@ project:
 bibliography: 
   - bib/book.bib
   - bib/packages.bib
+  - bib/blog.bib
```
[the remark this answers is not in the document]
Comment (christophscheuch) [reply _quarto.yml:9]: So we can have just links in the text vs one blog bib vs many blog.bibs per post, right? I'm fine with just providing links or footnotes on the blog page and not introducing a blog.bib.

File: blog/tidy-collaborative-filtering/thumbnail.png
[the remark this answers is not in the document]
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/thumbnail.png:1]: "DALL-E 2 users are allowed to use generated images for commercial purposes like printing, selling or licensing". Afaik all images also have a watermark, so we are safe.

File: blog/tidy-collaborative-filtering/index.qmd [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+title: "Tidy Collaborative Filtering: Building A Stock Recommender"
+author: "Christoph Scheuch"
+date: "2023-05-17"
+description: A simple implementation for prototyping multiple collaborative filtering algorithms.
+image: thumbnail.png
+categories: 
+  - Recommender System
+---
+
+Recommender systems are a key component of our digital lifes, ranging from e-commerce, online advertisements, movie recommendations, or more generally all kinds of product recommendations. A recommender system aims to efficiently deliver personalized content to users based on a large pool of potentially relevant information. In this blog post, I illustrate the concept of recommender systems in building a simple stock recommendation tool that relies on publicly available portfolios from the social trading platform [wikifolio.com](https://wikifolio.com/). The resulting recommender proposes stocks to investors who already have their own portfolio and look for new investment opportunities. The underlying assumption is that the wikifolio traders hold stock portfolios that can provide meaningful inspiration for other investors.
```
[the remark this answers is not in the document]
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:11]: I'll add it to be on the safe side.

File: blog/tidy-collaborative-filtering/thumbnail.png
[the remark this answers is not in the document]
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/thumbnail.png:1]: I added the info to the image-alt text.

File: blog/tidy-collaborative-filtering/index.qmd (hunk 10 above)
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:128]: plural because a list of numbers follows

File: bib/blog.bib [hunk 17]
```diff
@@ -0,0 +1,19 @@
+@inproceedings{BreeseEtAl1998,
```
[the remark this answers is not in the document]
Comment (christophscheuch) [reply bib/blog.bib:1]: Redudant since I replaced blog.bib references with footnotes.

File: _quarto.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (christophscheuch) [reply _quarto.yml:9]: I also just used footnotes, it is easier to maintain.

File: blog/tidy-collaborative-filtering/index.qmd (hunk 9 above)
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:85]: They / them works better in this case

File: blog/tidy-collaborative-filtering/index.qmd [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+
+## Content-Based Filering
+
+Content-based filtering methods exploit information about users or items to create recommendations by building a model that relates available characteristics of users or items to each other. The recommendation problem is hence cast into a classification problem (the user will like the item or not) or more generally a regression problem (which rating will the user give an item). The classification problem can be item-centered by focusing on available user information and estimating a model for each item. If there are a lot of user-item interactions available, the resulting model is fairly robust, but it is less personalized (as it ignores user characteristics apart from interactions). The classification problem can also be user-centered by working with item features and estimating a model for each user. However, if a user only has a few interactions then the resulting model becomes easily unstable. Content-based filtering can also be neither the user nor item-centered by stacking the two feature vectors and putting them into a neural network.
```
[the remark this answers is not in the document]
Comment (christophscheuch) [reply blog/tidy-collaborative-filtering/index.qmd:28]: Content-based filtering can also be neither user nor item-centered by stacking the two feature vectors, hence considering both input simultaneously, and putting them into a neural network
